File: src/WorkBC.Shared/Services/GeocodingService.cs
using System;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using System.Xml.Linq;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using WorkBC.Data.Model.JobBoard;
using WorkBC.Shared.Settings;

namespace WorkBC.Shared.Services
{
    public class GeocodingService : IGeocodingApiService
    {
        private readonly ILogger<IGeocodingService> _logger;
        private readonly IConfiguration _configuration;

        public GeocodingService(IConfiguration configuration,
            ILogger<IGeocodingService> logger = null)
        {
            _logger = logger;
            _configuration = configuration;
        }

        public async Task<GeocodedLocationCache> GetLocation(string location)
        {
            //Get the long and lat of the current location
            (string lat, string lon, string city, string frenchCity, string province) geo =
                await GetGoogleMapsLocation(location);

            if (geo.lat != null && geo.lon != null)
            {
                var glc = new GeocodedLocationCache
                {
                    DateGeocoded = DateTime.Now,
                    Name = location,
                    Latitude = geo.lat,
                    Longitude = geo.lon,
                    City = geo.city,
                    FrenchCity = geo.frenchCity,
                    Province = geo.province
                };
                return glc;
            }

            return null;
        }

        private async Task<(string lat, string lon, string city, string frenchCity, string province)>
            GetGoogleMapsLocation(string address)
        {
            XDocument xdoc = null;

            try
            {
                //read API key
                string apiKey = _configuration["AppSettings:GoogleMapsIPApi"];

                var requestUri =
                    $"https://maps.googleapis.com/maps/api/geocode/xml?address={Uri.EscapeDataString(address)}&key={apiKey}";

                //Read into XDocument
                xdoc = await GetWebResponse(requestUri);

                //Get the result node
                XElement result = xdoc.Element("GeocodeResponse")?.Element("result");

                if (result == null)
                {
                    return (null, null, null, null, null);
                }

                //Parse the city
                string city = GetCity(result);

                // Get the province
                string province = result.Elements("address_component")
                    ?.Where(x => x.Element("type")?.Value == "administrative_area_level_1")
                    ?.SingleOrDefault()
                    ?.Element("short_name")
                    ?.Value;

                //Get the location node
                XElement locationElement = result?.Element("geometry")?.Element("location");

                //Read the long and lat from the XML
                XElement lat = locationElement?.Element("lat");
                XElement lng = locationElement?.Element("lng");

                string frenchCity = null;

                // only look up the French city if the address contains a number
                // (because we only really use it for postal codes)
                if (address.Any(char.IsDigit)) 
                {
                    //Get the french version
                    XDocument frenchXDoc = await GetWebResponse(requestUri, true);

                    //Get the french result node
                    XElement frenchResult = frenchXDoc.Element("GeocodeResponse")?.Element("result");

                    //Parse the french city
                    frenchCity = GetCity(frenchResult);
                }

                //Return long and lat
                return (lat?.Value, lng?.Value, city, frenchCity, province);
            }
            catch (Exception ex)
            {
                if (_logger != null)
                {
                    _logger.LogError("GetGoogleMapsLocation(string address, IConfiguration configuration) : " +
                                     ex.Message);

                    //Most likely a problem with the Google API
                    //More details will be in the response from the API
                    if (xdoc != null)
                    {
                        _logger.LogError("Google API Response : " + xdoc);
                    }
                }
            }

            //Error
            return (null, null, null, null, null);
        }

        private static string GetCity(XElement result)
        {
            string city = result?.Elements("address_component")
                ?.Where(x => x.Element("type")?.Value == "locality")
                ?.SingleOrDefault()
                ?.Element("short_name")
                ?.Value;

            if (city == null)
            {
                city = result?.Elements("address_component")
                    ?.Where(x => x.Element("type")?.Value == "neighborhood")
                    ?.SingleOrDefault()
                    ?.Element("short_name")
                    ?.Value;
            }

            return city;
        }

        /// <summary>
        ///     Get XML data from URL response
        /// </summary>
        private async Task<XDocument> GetWebResponse(string url, bool isFrench = false)
        {
            var proxySettings = new ProxySettings();
            _configuration.GetSection("ProxySettings").Bind(proxySettings);

            try
            {
                //response from server
                string responseFromServer;

                var handler = new HttpClientHandler();

                if (proxySettings.UseProxy)
                {
                    handler.Proxy = new WebProxy(proxySettings.ProxyHost, proxySettings.ProxyPort)
                    {
                        BypassProxyOnLocal = true
                    };
                }

                if (proxySettings.IgnoreSslErrors)
                {
                    handler.ClientCertificateOptions = ClientCertificateOption.Manual;
                    handler.ServerCertificateCustomValidationCallback =
                        (httpRequestMessage, cert, cetChain, policyErrors) => true;
                }

                //Create new web request to URL
                using (var httpClient = new HttpClient(handler))
                {
                    if (isFrench)
                    {
                        url += "&language=fr";
                    }

                    //Read the web response from URL
                    HttpResponseMessage response = await httpClient.GetAsync(url);

                    //Save response
                    responseFromServer = await response.Content.ReadAsStringAsync();
                }

                //load web request to xml
                return XDocument.Parse(responseFromServer);
            }
            catch
            {
                return null;
            }
        }
    }
}

File: src/WorkBC.Notifications.JobAlerts.V2/src/Search/GeocodingService.php
<?php

declare(strict_types=1);

namespace WorkBC\JobAlertNotifier\Search;

use GuzzleHttp\Client;
use Monolog\Logger;
use WorkBC\JobAlertNotifier\Config\AppConfig;

/**
 * Port of WorkBC.Shared.Services.GeocodingService.
 *
 * Looks a location key (e.g. "V6B4N7, CANADA" or "Surrey, BC, CANADA") up in
 * the GeocodedLocationCache table; on a miss it geocodes via the Google Maps
 * XML API and stores the result. Latitude/Longitude are varchar columns and
 * are concatenated verbatim into the ES query, exactly like the C#.
 */
final class GeocodingService implements GeocodingInterface
{
    private Client $http;

    public function __construct(
        private readonly \PDO $pdo,
        private readonly AppConfig $config,
        private readonly Logger $log,
    ) {
        $options = [
            'timeout' => $config->httpTimeout,
            'http_errors' => false,
        ];
        if (($proxy = $config->outboundProxy()) !== null) {
            $options['proxy'] = $proxy;
        }
        if ($config->proxyIgnoreSslErrors) {
            $options['verify'] = false;
        }
        $this->http = new Client($options);
    }

    /**
     * @return array{Latitude: string, Longitude: string}|null
     */
    public function getLocation(string $location): ?array
    {
        $stmt = $this->pdo->prepare(
            'SELECT "Latitude", "Longitude" FROM "GeocodedLocationCache" WHERE "Name" = :name LIMIT 1'
        );
        $stmt->execute([':name' => $location]);
        $row = $stmt->fetch();

        if ($row !== false) {
            return ['Latitude' => (string) $row['Latitude'], 'Longitude' => (string) $row['Longitude']];
        }

        return $this->createLocation($location);
    }

    /**
     * @return array{Latitude: string, Longitude: string}|null
     */
    private function createLocation(string $location): ?array
    {
        $geo = $this->getGoogleMapsLocation($location);

        if ($geo['lat'] !== null && $geo['lon'] !== null) {
            $stmt = $this->pdo->prepare(
                'INSERT INTO "GeocodedLocationCache"
                    ("Name", "Latitude", "Longitude", "City", "FrenchCity", "Province", "DateGeocoded", "IsPermanent")
                 VALUES (:name, :lat, :lon, :city, :frenchCity, :province, NOW(), FALSE)'
            );
            $stmt->execute([
                ':name' => $location,
                ':lat' => $geo['lat'],
                ':lon' => $geo['lon'],
                ':city' => $geo['city'],
                ':frenchCity' => $geo['frenchCity'],
                ':province' => $geo['province'],
            ]);

            return ['Latitude' => $geo['lat'], 'Longitude' => $geo['lon']];
        }

        return null;
    }

    /**
     * @return array{lat: ?string, lon: ?string, city: ?string, frenchCity: ?string, province: ?string}
     */
    private function getGoogleMapsLocation(string $address): array
    {
        $none = ['lat' => null, 'lon' => null, 'city' => null, 'frenchCity' => null, 'province' => null];

        try {
            $requestUri = 'https://maps.googleapis.com/maps/api/geocode/xml?address='
                . rawurlencode($address) . '&key=' . $this->config->googleMapsApiKey;

            $xdoc = $this->getWebResponse($requestUri);
            $result = $xdoc?->result;

            if ($result === null || count($result) === 0) {
                return $none;
            }
            $result = $result[0];

            $city = $this->getCity($result);
            $province = $this->getAddressComponent($result, 'administrative_area_level_1');

            $lat = isset($result->geometry->location->lat) ? (string) $result->geometry->location->lat : null;
            $lon = isset($result->geometry->location->lng) ? (string) $result->geometry->location->lng : null;

            $frenchCity = null;

            // only look up the French city if the address contains a number
            // (because we only really use it for postal codes)
            if (preg_match('/\d/', $address) === 1) {
                $frenchDoc = $this->getWebResponse($requestUri, true);
                $frenchResult = ($frenchDoc?->result ?? null);
                if ($frenchResult !== null && count($frenchResult) > 0) {
                    $frenchCity = $this->getCity($frenchResult[0]);
                }
            }

            return ['lat' => $lat, 'lon' => $lon, 'city' => $city, 'frenchCity' => $frenchCity, 'province' => $province];
        } catch (\Throwable $e) {
            $this->log->error('getGoogleMapsLocation(' . $address . ') : ' . $e->getMessage());
        }

        return $none;
    }

    private function getCity(\SimpleXMLElement $result): ?string
    {
        return $this->getAddressComponent($result, 'locality')
            ?? $this->getAddressComponent($result, 'neighborhood');
    }

    private function getAddressComponent(\SimpleXMLElement $result, string $type): ?string
    {
        foreach ($result->address_component as $component) {
            foreach ($component->type as $componentType) {
                if ((string) $componentType === $type) {
                    return isset($component->short_name) ? (string) $component->short_name : null;
                }
            }
        }
        return null;
    }

    private function getWebResponse(string $url, bool $isFrench = false): ?\SimpleXMLElement
    {
        try {
            if ($isFrench) {
                $url .= '&language=fr';
            }

            $response = $this->http->get($url);
            $body = (string) $response->getBody();

            $previous = libxml_use_internal_errors(true);
            $xml = simplexml_load_string($body);
            libxml_use_internal_errors($previous);

            return $xml === false ? null : $xml;
        } catch (\Throwable) {
            return null;
        }
    }
}
